### PR TITLE
Fixed #543 : Added a feature: Alt + <digit> to switch between tabs.

### LINKED ISSUE #543: Feature: Alt + <digit> to switch between tabs
It would be nice to all a possibility to switch between opened tabs by
ALT + <digit> hotkeys, like it work in Firefox and Chrome.

For example, Alt+1 activates the first opened tab, Alt+2 activates the second, and so on.
Alt+0 activates the 10-th tab.

### PR DESCRIPTION
**Pull request by rishabh-soni**
Resolved #543 - I made different Actions for each individual key sequence (Alt + 1, Alt +2,..) namely ToggleTab1Action, ToggleTab2Action.. and so on.

### CODE REVIEW

**Comment by asashnov** [comment]
A better approach could be
* creating a subclass from QAction, for example, `SetCurrentTabAction`;
* Passing name, hotkey, index (0-9), and a pointer to the main window to the constructor;
*  we can make those actions 'hidden' to not show them in Menu;
* in receiving slot, we can use  qobject_cast<SetCurrentTabAction*>(sender())  and if this is not nullptr, get the needed tab index from it.

**Comment by kelson42** [comment]
@rishabh-soni Any update?

**Comment by rishabh-soni** [comment]
Yeah I am working on it 👍

**Comment by asashnov** [comment]
Just realized the more easy approach: we can use QAction::data() with the tab number

**Comment by kelson42** [comment]
@rishabh-soni Hopefuly the last comment by @asashnov will help you to complete this PR. Otherwise please let us know.

**Comment by kelson42** [comment]
@rishabh-soni Any news here? We plan to do a release soon and it would be great to have your PR inside.

**Comment by rishabh-soni** [comment]
Oh I see, actually my mid semester exams have begun and the last exam is on 1st Feb. So can you give me a deadline or time till 2nd or 3rd Feb?

**Comment by kelson42** [comment]
3rd of february is OK.

**Comment by rishabh-soni** [comment]
I am trying to implement the sub-classing QAction approach as said, but I am new to this framework, so I am having a hard time figuring out how would I extract (I am not sure if that would be the right word to say here) the number which was passed along with the 'alt' key? Can anyone please give me a hint or a source from which I can see some example of a similar implementation?

**Comment by kelson42** [comment]
@asashnov ?

**Comment by asashnov** [comment]
@rishabh-soni  No need to subclass `QAction`, just use `QAction::data()`. Here is an example:

```c
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    for (int i = 0 ; i < ui->tabWidget->count() ; i++) {
        QAction *a = new QAction(this);
        a->setText(QString("Switch to tab %1").arg(i+1));
        a->setData(QVariant::fromValue(i));
        QKeySequence ks(Qt::ALT | (Qt::Key_1 + i));
        a->setShortcut(ks);
        addAction(a);
        connect(a, SIGNAL(triggered()), this, SLOT(switch_to_tab_by_triggered_action()));
    }
}

void MainWindow::switch_to_tab_by_triggered_action()
{
    QAction *a = qobject_cast<QAction*>(QObject::sender());
    if (!a)
        return;

    bool ok;
    int tab_n = a->data().toInt(&ok);

    if (!ok)
        return;
    if (tab_n >= ui->tabWidget->count())
        return;

    ui->tabWidget->setCurrentIndex(tab_n);
}
```

To run this example, just create a new Qt Widget Application project in QtCreator, put QTabWidget into MainWindow, and add this code. The whole project is here:

[SwitchingTabs.zip](https://github.com/kiwix/kiwix-desktop/files/6043833/SwitchingTabs.zip)

File: src/kiwixapp.cpp [hunk 1]
```diff
@@ -307,6 +307,36 @@ void KiwixApp::createAction()
...
+    SET_SHORTCUT(ToggleTab4Action, QKeySequence(Qt::ALT + Qt::Key_4));
+
+    CREATE_ACTION(ToggleTab5Action, gt("toggle-tab-5"));
+    SET_SHORTCUT(ToggleTab5Action, QKeySequence(Qt::ALT + Qt::Key_5));
+
+    CREATE_ACTION(ToggleTab6Action, gt("toggle-tab-6"));
+    SET_SHORTCUT(ToggleTab6Action, QKeySequence(Qt::ALT + Qt::Key_6));
+
+    CREATE_ACTION(ToggleTab7Action, gt("toggle-tab-7"));
+    SET_SHORTCUT(ToggleTab7Action, QKeySequence(Qt::ALT + Qt::Key_7));
```
Comment (asashnov) [inline src/kiwixapp.cpp:329]: Maybe Qt::AltModifier | Qt::Key_7  is a better option, according to
 https://doc.qt.io/qt-5/qt.html#KeyboardModifier-enum

**Review by asashnov — COMMENTED** [review]
Probably we don't have to use CREATE_ACTION macros for Alt+<N> shortcuts.

**Comment by rishabh-soni** [comment]
Thanks a lot, I'll implement this as soon as possible 👍🏼

**Comment by kelson42** [comment]
@rishabh-soni any news?

**Comment by rishabh-soni** [comment]
Done! Thanks a lot @asashnov, now I have made the required changes using QAction::data(). Please have a look now at my commit.

**Comment by kelson42** [comment]
@rishabh-soni Great, so this is not a draft PR anymore?

**Comment by kelson42** [comment]
@rishabh-soni You could i believe rebase your branch on git origin master, that would be better.

File: kiwix-desktop.pro.user [hunk 2]
```diff
@@ -0,0 +1,320 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (asashnov) [inline kiwix-desktop.pro.user:1]: .pro.user file is local for every developer's machine, no need to commit it to repository

File: src/mainmenu.cpp [hunk 3]
```diff
@@ -11,7 +11,6 @@ MainMenu::MainMenu(QWidget *parent) :
 {
     auto app = KiwixApp::instance();
     setStyle(new MenuProxyStyle());
-
```
Comment (asashnov) [inline src/mainmenu.cpp:14]: I like this extra line to logically separate blocks of code. As we don't modify any code in this file, let's just exclude it from this PR.

File: src/tabbar.h [hunk 4]
```diff
@@ -65,7 +65,6 @@ public slots:
     QStackedWidget*     mp_stackedWidget;
     int                 m_settingsIndex;
     QScopedPointer<FullScreenWindow> m_fullScreenWindow;
-
```
Comment (asashnov) [inline src/tabbar.h:68]: This empty line visually separates properties definition from methods. So let's keep it.

File: src/tabbar.cpp [hunk 5]
```diff
@@ -26,42 +26,66 @@ TabBar::TabBar(QWidget *parent) :
     auto app = KiwixApp::instance();
     connect(app->getAction(KiwixApp::NewTabAction), &QAction::triggered,
             this, [=]() {
-                this->createNewTab(true);
```
Comment (asashnov) [inline src/tabbar.cpp:29]: @mgautierfr  Do we have any code convention on Kiwix Desktop project? If not, I suggest using Qt's convention: https://wiki.qt.io/Coding_Conventions#Lambdas

**Review by asashnov — CHANGES_REQUESTED** [review]

**Comment by kelson42** [comment]
@rishabh-soni Please reopen the PR... and be carefuly with force pushing, you might loose your code if you don't take care.

**Comment by rishabh-soni** [comment]
Yeah just give me some time I had made a backup of my code so I'll restore it shortly.

**Comment by rishabh-soni** [comment]
@asashnov @kelson42 Please have a look now, I have updated my fork and all the spaces and empty lines which were previously in the code are restored now. By mistake I force pushed the re based branch into my fork.

**Review by asashnov — APPROVED** [review]
Looks good and works, well done!

**Pull request merged**